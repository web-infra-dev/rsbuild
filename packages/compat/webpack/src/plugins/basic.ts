import path from 'path';
import type { RsbuildPlugin } from '../types';
import { applyBasicPlugin } from '@rsbuild/shared';

/**
 * Provide some basic configs of webpack
 */
export const pluginBasic = (): RsbuildPlugin => ({
  name: 'plugin-basic',

  setup(api) {
    applyBasicPlugin(api);

    api.modifyWebpackChain(async (chain, { env }) => {
      // Disable webpack performance hints.
      // These logs are too complex
      chain.performance.hints(false);

      // This will be futureDefaults in webpack 6
      chain.module.parser.merge({
        javascript: {
          exportsPresence: 'error',
        },
      });

      if (env === 'development') {
        // Set correct path for source map
        // this helps VS Code break points working correctly in monorepo
        chain.output.devtoolModuleFilenameTemplate(
          (info: { absoluteResourcePath: string }) =>
            path.resolve(info.absoluteResourcePath).replace(/\\/g, '/'),
        );
      }
    });
  },
});
