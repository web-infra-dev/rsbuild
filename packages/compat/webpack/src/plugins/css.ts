import {
  CSS_REGEX,
  resolvePackage,
  isUseCssExtract,
  getPostcssConfig,
  ModifyChainUtils,
  isUseCssSourceMap,
  mergeChainedOptions,
  getCssLoaderOptions,
  getBrowserslistWithDefault,
  getCssModuleLocalIdentName,
  type Context,
  type BundlerChainRule,
} from '@rsbuild/shared';
import type {
  RsbuildPlugin,
  CSSExtractOptions,
  NormalizedConfig,
} from '../types';

export async function applyBaseCSSRule({
  rule,
  config,
  context,
  utils: { target, isProd, isServer, CHAIN_ID, isWebWorker, getCompiledPath },
  importLoaders = 1,
}: {
  rule: BundlerChainRule;
  config: NormalizedConfig;
  context: Context;
  utils: ModifyChainUtils;
  importLoaders?: number;
}) {
  const browserslist = await getBrowserslistWithDefault(
    context.rootPath,
    config,
    target,
  );

  // 1. Check user config
  const enableExtractCSS = isUseCssExtract(config, target);
  const enableSourceMap = isUseCssSourceMap(config);
  const enableCSSModuleTS = Boolean(config.output.enableCssModuleTSDeclaration);

  // 2. Prepare loader options
  const localIdentName = getCssModuleLocalIdentName(config, isProd);

  const cssLoaderOptions = getCssLoaderOptions({
    config,
    enableSourceMap,
    importLoaders,
    isServer,
    isWebWorker,
    localIdentName,
  });

  // 3. Create webpack rule
  // Order: style-loader/mini-css-extract -> css-loader -> postcss-loader
  if (!isServer && !isWebWorker) {
    // use mini-css-extract-plugin loader
    if (enableExtractCSS) {
      const extraCSSOptions: Required<CSSExtractOptions> =
        typeof config.tools.cssExtract === 'object'
          ? config.tools.cssExtract
          : {
              loaderOptions: {},
              pluginOptions: {},
            };

      rule
        .use(CHAIN_ID.USE.MINI_CSS_EXTRACT)
        .loader(require('mini-css-extract-plugin').loader)
        .options(extraCSSOptions.loaderOptions)
        .end();
    }
    // use style-loader
    else {
      const styleLoaderOptions = mergeChainedOptions({
        defaults: {},
        options: config.tools.styleLoader,
      });

      rule
        .use(CHAIN_ID.USE.STYLE)
        .loader(require.resolve('style-loader'))
        .options(styleLoaderOptions)
        .end();
    }

    // use css-modules-typescript-loader
    if (enableCSSModuleTS && cssLoaderOptions.modules) {
      rule
        .use(CHAIN_ID.USE.CSS_MODULES_TS)
        .loader(
          resolvePackage(
            '@rsbuild/shared/css-modules-typescript-loader',
            __dirname,
          ),
        )
        .options({
          modules: cssLoaderOptions.modules,
        })
        .end();
    }
  } else {
    rule
      .use(CHAIN_ID.USE.IGNORE_CSS)
      .loader(resolvePackage('@rsbuild/shared/ignore-css-loader', __dirname))
      .end();
  }

  rule
    .use(CHAIN_ID.USE.CSS)
    .loader(getCompiledPath('css-loader'))
    .options(cssLoaderOptions)
    .end();

  if (!isServer && !isWebWorker) {
    const postcssLoaderOptions = getPostcssConfig({
      enableSourceMap,
      browserslist,
      config,
    });

    rule
      .use(CHAIN_ID.USE.POSTCSS)
      .loader(getCompiledPath('postcss-loader'))
      .options(postcssLoaderOptions)
      .end();
  }

  // CSS imports should always be treated as sideEffects
  rule.merge({ sideEffects: true });

  // Enable preferRelative by default, which is consistent with the default behavior of css-loader
  // see: https://github.com/webpack-contrib/css-loader/blob/579fc13/src/plugins/postcss-import-parser.js#L234
  rule.resolve.preferRelative(true);
}

export const pluginCss = (): RsbuildPlugin => {
  return {
    name: 'plugin-css',
    setup(api) {
      api.modifyBundlerChain(async (chain, utils) => {
        const rule = chain.module.rule(utils.CHAIN_ID.RULE.CSS);
        const config = api.getNormalizedConfig();
        rule.test(CSS_REGEX);
        await applyBaseCSSRule({
          rule,
          utils,
          config,
          context: api.context,
        });
      });
    },
  };
};
