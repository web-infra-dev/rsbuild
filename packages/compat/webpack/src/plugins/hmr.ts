import { isUsingHMR, type RsbuildPlugin } from '@rsbuild/shared';

export const pluginHMR = (): RsbuildPlugin => ({
  name: 'rsbuild-webpack:hmr',

  setup(api) {
    api.modifyBundlerChain((chain, utils) => {
      const config = api.getNormalizedConfig();

      if (!isUsingHMR(config, utils)) {
        return;
      }

      const { bundler, CHAIN_ID } = utils;
      chain.plugin(CHAIN_ID.PLUGIN.HMR).use(bundler.HotModuleReplacementPlugin);
    });
  },
});
