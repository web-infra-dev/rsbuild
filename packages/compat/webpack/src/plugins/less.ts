import {
  isUseCssSourceMap,
  LESS_REGEX,
  FileFilterUtil,
  getLessLoaderOptions,
} from '@rsbuild/shared';
import type { RsbuildPlugin } from '../types';

export type LessLoaderUtils = {
  addExcludes: FileFilterUtil;
};

export function pluginLess(): RsbuildPlugin {
  return {
    name: 'rsbuild-webpack:less',
    setup(api) {
      api.modifyBundlerChain(async (chain, utils) => {
        const config = api.getNormalizedConfig();
        const { applyBaseCSSRule } = await import('./css');

        const { options, excludes } = getLessLoaderOptions(
          config.tools.less,
          isUseCssSourceMap(config),
        );
        const rule = chain.module
          .rule(utils.CHAIN_ID.RULE.LESS)
          .test(LESS_REGEX);

        excludes.forEach((item) => {
          rule.exclude.add(item);
        });

        await applyBaseCSSRule({
          rule,
          utils,
          config,
          context: api.context,
          importLoaders: 2,
        });

        rule
          .use(utils.CHAIN_ID.USE.LESS)
          .loader(utils.getCompiledPath('less-loader'))
          .options(options);
      });
    },
  };
}
