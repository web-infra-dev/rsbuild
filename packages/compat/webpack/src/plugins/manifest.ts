import type { RsbuildPlugin } from '../types';
import { generateManifest } from '@rsbuild/shared';

export const pluginManifest = (): RsbuildPlugin => ({
  name: 'plugin-manifest',

  setup(api) {
    api.modifyWebpackChain(async (chain, { CHAIN_ID }) => {
      const config = api.getNormalizedConfig();

      if (!config.output.enableAssetManifest) {
        return;
      }

      const { WebpackManifestPlugin } = await import(
        '../../compiled/webpack-manifest-plugin'
      );
      const publicPath = chain.output.get('publicPath');

      chain.plugin(CHAIN_ID.PLUGIN.MANIFEST).use(WebpackManifestPlugin, [
        {
          fileName: 'asset-manifest.json',
          publicPath,
          generate: generateManifest,
        },
      ]);
    });
  },
});
