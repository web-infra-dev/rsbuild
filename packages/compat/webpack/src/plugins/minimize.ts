import {
  CHAIN_ID,
  BundlerChain,
  getJSMinifyOptions,
  type RsbuildPlugin,
  type NormalizedConfig,
} from '@rsbuild/shared';

async function applyJSMinimizer(chain: BundlerChain, config: NormalizedConfig) {
  const { default: TerserPlugin } = await import('terser-webpack-plugin');

  const finalOptions = await getJSMinifyOptions(config);

  chain.optimization
    .minimizer(CHAIN_ID.MINIMIZER.JS)
    .use(TerserPlugin, [
      // Due to terser-webpack-plugin has changed the type of class, which using a generic type in
      // constructor, leading auto inference of parameters of plugin constructor is not possible, using any instead
      finalOptions as any,
    ])
    .end();
}

export const pluginMinimize = (): RsbuildPlugin => ({
  name: 'rsbuild-webpack:minimize',

  setup(api) {
    api.modifyBundlerChain(async (chain, { isProd }) => {
      const config = api.getNormalizedConfig();
      const isMinimize = isProd && !config.output.disableMinimize;

      // set minimize to allow users to disable minimize
      chain.optimization.minimize(isMinimize);

      if (isMinimize) {
        await applyJSMinimizer(chain, config);
      }
    });
  },
});
