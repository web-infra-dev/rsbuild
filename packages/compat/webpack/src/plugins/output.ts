import { posix } from 'path';
import {
  getDistPath,
  getFilename,
  isUseCssExtract,
  applyOutputPlugin,
  mergeChainedOptions,
  type RsbuildPlugin,
} from '@rsbuild/shared';

export const pluginOutput = (): RsbuildPlugin => ({
  name: 'rsbuild-webpack:output',

  setup(api) {
    applyOutputPlugin(api);

    api.modifyBundlerChain(async (chain, { isProd, target, CHAIN_ID }) => {
      const config = api.getNormalizedConfig();

      const cssPath = getDistPath(config.output, 'css');

      // css output
      if (isUseCssExtract(config, target)) {
        const { default: MiniCssExtractPlugin } = await import(
          'mini-css-extract-plugin'
        );
        const extractPluginOptions = mergeChainedOptions({
          defaults: {},
          options: config.tools.cssExtract?.pluginOptions,
        });

        const cssFilename = getFilename(config.output, 'css', isProd);

        chain
          .plugin(CHAIN_ID.PLUGIN.MINI_CSS_EXTRACT)
          .use(MiniCssExtractPlugin, [
            {
              filename: posix.join(cssPath, cssFilename),
              chunkFilename: posix.join(cssPath, `async/${cssFilename}`),
              ignoreOrder: true,
              ...extractPluginOptions,
            },
          ]);
      }
    });
  },
});
