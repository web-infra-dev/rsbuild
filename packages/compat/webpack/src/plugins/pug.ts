import path from 'path';
import type { RsbuildPlugin } from '../types';
import { mergeChainedOptions } from '@rsbuild/shared';

export const pluginPug = (): RsbuildPlugin => ({
  name: 'plugin-pug',

  setup(api) {
    api.modifyWebpackChain(async (chain, { CHAIN_ID }) => {
      const config = api.getNormalizedConfig();
      const { pug } = config.tools;
      if (!pug) {
        return;
      }

      chain.module
        .rule(CHAIN_ID.RULE.PUG)
        .test(/\.pug$/)
        .use(CHAIN_ID.USE.PUG)
        .loader(path.resolve(__dirname, '../webpackLoaders/pugLoader'))
        .options(mergeChainedOptions({}, pug === true ? {} : pug));
    });
  },
});
