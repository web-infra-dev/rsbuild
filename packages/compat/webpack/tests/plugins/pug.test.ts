import { pluginPug } from '@/plugins/pug';
import { createStubRsbuild } from '../helper';

describe('plugin-pug', () => {
  it('should add pug rules correctly when tools.pug is used', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginPug()],
      rsbuildConfig: {
        tools: {
          pug: {
            pretty: true,
          },
        },
      },
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should not add pug rules correctly when tools.pug is not used', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginPug()],
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config).toEqual({});
  });
});
