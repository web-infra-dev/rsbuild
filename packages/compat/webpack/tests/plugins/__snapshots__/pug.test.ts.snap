// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-pug > should add pug rules correctly when tools.pug is used 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.pug\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/compat/webpack/src/webpackLoaders/pugLoader",
            "options": {
              "pretty": true,
            },
          },
        ],
      },
    ],
  },
}
`;
